### ISSUE TITLE
Powerline Glyph for Git Branch

### ISSUE THREAD

**Issue by krish-r** [arrowtype/recursive#351]
Hi, 
Could you please consider adding the powerline glyph for git branch if possible?  

OS:                        Windows 10 
Terminal:                VS Code Integrated terminal
Recursive Version:  v1.051

Cascadia Code: 
![image](https://user-images.githubusercontent.com/54745129/84574480-4e797780-adc4-11ea-80d7-745f6ff31d3a.png)

Recursive Mono:
![image](https://user-images.githubusercontent.com/54745129/84574525-a57f4c80-adc4-11ea-994c-d6bd9f92eadb.png)

Thanks

**Comment by arrowtype** [comment]
This is a good idea, thanks!

It looks like the Windows Terminal doesn’t yet support fallback fonts, but maybe it would be helpful to upvote/add a reaction to it at https://github.com/microsoft/terminal/issues/1790 to encourage that effort. :)

**Comment by arrowtype** [comment]
Note for later: seems like https://github.com/ryanoasis/nerd-fonts might be a way to do this

**Comment by arrowtype** [comment]
Notes:

There are fewer powerline glyphs that I realized – I thought they might include a lot of box-drawing glyphs, but it is basically 8 useful glyphs, then a bunch of glyphs that seem to be purely for ... style?

From [nerd fonts powerline fonts](https://github.com/ryanoasis/nerd-fonts/tree/c41890f82b4d346cbfffff25ae29d1e145690d13/src/glyphs), here is the basic set:

![image](https://user-images.githubusercontent.com/45946693/93283689-0856fa00-f79f-11ea-8aae-e77ec1574401.png)

...and the "extras":

![image](https://user-images.githubusercontent.com/45946693/93283725-1e64ba80-f79f-11ea-8701-c845a38e933a.png)

A screenshot from the powerline website shows some of these in context, as elements that are seemingly just for decoration (rather than system status, etc):

![image](https://user-images.githubusercontent.com/45946693/93283564-c2019b00-f79e-11ea-93cc-e20866307f50.png)

From Cascadia Code, here is the glyph set, adapted to fit within single-width units:

![image](https://user-images.githubusercontent.com/45946693/93283585-d04fb700-f79e-11ea-8a33-8ee47f48974c.png)

If/when I add these to Recursive, I would have a few goals:
- Add them just at the Regular weight
- Possibly adapt their style slightly to match Recursive, and maaaaaybe even give them a Casual axis
- Match their overall sizing to work well with Recursive
- Probably, have them added in a post-process to the Rec Mono Code fonts, rather than being included in the main fonts

**Comment by arrowtype** [comment]
Design steps:
0. Make a branch, e.g. `powerline`
1. Start with the [Cascadia Code powerline glyphs](https://github.com/microsoft/cascadia-code/tree/716c6e6f24ce8e23e28936db9a106a6129ca9931/sources/nerdfonts)
2. Adjust this to a 1000 UPM, and glyphs to 600 units wide
3. Use scaling transforms to bring the glyphs to the correct height (probably, figure our the percentage transform needed, then transform all glyphs)
4. Break into three fonts: `powerline-extras.ufo`, `powerline-base-casual.ufo`, and `powerline-base-linear.ufo`
5. Adjust "Casual" and "Linear" versions of the main glyphs:
  - Make the "branch" glyph have a Recursive-ish arrow, but keep it connected because it’s too small for much detail
  - The `LN` and `CN` glyphs can get adjusted Recursive caps. Probably, the "C" should be more like the lowercase shape, to keep an open shape (but not be the rectangular thing typically in PL glyphs)
  - Make the "lock" glyph not ugly, if possible (hard to say what should change, though ... probably, it should just get Recursive-style corners and outlines)
6. In the `powerline-extras`, make an even amount of space in contour lines – e.g. the legos should have even thicknesses in their negative spaces, rather than looking squished (they currently look squished in Cascadia, as shown below)

![image](https://user-images.githubusercontent.com/45946693/93285352-a4363500-f7a2-11ea-9ae0-ea7aef294513.png)

7. Make a Pull Request with that branch

**Comment by arrowtype** [comment]
We’ve made some good progress here with the PR https://github.com/arrowtype/recursive/pull/407.

Next, I will work on a way to add this into the build system for Rec Mono fonts (e.g. https://github.com/arrowtype/recursive-code-config).

**Comment by arrowtype** [comment]
Approach I’ll try here:
- export from powerline-specific UFOs to basic TTFs
- use the FontTools Merger to merge those with the code TTFs

**Comment by arrowtype** [comment]
This is partly working! I have a starter script here: https://github.com/arrowtype/recursive-code-config/blob/5260e83be9669e6e6cce4ea1e89510af6d11462f/scripts/merge-powerline-ttfs.py

However, there is a funny (slightly stupid on my part) mistake preventing the "branch" glyph from display with Recursive: the glyph already exists in the fonts, but only as a blank glyph (basically like a word space). This is probably because I added the glyph earlier, with the intention of filling it with an Recursive-specific branch drawing.

However, this results in that icon showing up as a blank space, if Recursive is used with a PowerLine theme. 💀

![image](https://user-images.githubusercontent.com/45946693/94886827-8b847b00-0442-11eb-944c-d95fa9d772ac.png)

The fix is simple but will take a bit of time to allow for a new build & release: 
- I will remove the empty `uniE0A0` in the Recursive sources, then build a new version of the variable font.
- I will generate new RecMono Code fonts
- I will then improve the merger script in recursive-code-config to merge the powerline glyphs with all of the font styles, splitting linear vs casual.

I think this might be all it takes! 🤞 Or, it might take more investigation.

**Comment by arrowtype** [comment]
Got it working at a simple level!

![image](https://user-images.githubusercontent.com/45946693/94966481-5bcd8580-04cb-11eb-934d-a9015810a8b7.png)

I still need to:

- [ ] match the font metrics of the powerline UFOs & rebuild the TTFs
- [ ] then improve the merger script in recursive-code-config to merge the powerline glyphs with all of the font styles, splitting linear vs casual.

**Comment by arrowtype** [comment]
Design fixes still needed:

- [ ] adjust left side of “branch” glyph to balance with visual weight of right side

![image](https://user-images.githubusercontent.com/45946693/94993959-81f23480-0562-11eb-8098-2358688a5647.png)

- [ ] fix interior lines of these blocks `uniE0CC uniE0CD`:

![image](https://user-images.githubusercontent.com/45946693/94993898-22942480-0562-11eb-98b5-cf095441e9d3.png)

- [ ] give triangles little flat bits:

![image](https://user-images.githubusercontent.com/45946693/94993940-6850ed00-0562-11eb-8f56-c90b61b55d71.png)

**Comment by arrowtype** [comment]
Technical notes on the `.stypo` suffixes in Cascadia Code: https://twitter.com/ArrowType/status/1313594838462484481

TLDR: In , the `.stypo` glyphs match the typo metrics, and they are activated with the `rclt` OpenType feature.

Me: 

> I see that the Cascadia Code PowerLine UFO has both normal & "stypo" versions of PL glyphs.
> 
> 1. The stypo versions are probably for settings like the VScode terminal with shorter line heights, yes?
> 
> 2. How do you make that work?
> 
> Thank you for any insights!

Aaron Bell:

> Windows Terminal (and others) correctly implement the useTypoMetrics flag, so the idea of the sTypo versions is to adjust the heights so that they align with the sTypo metrics. 
> 
> For other systems that don’t respect that flag, they’d see the normal versions.

> Unfortunately, I’m not totally sure if it is actually working as intended (https://github.com/microsoft/cascadia-code/issues/310). It is complicated by the fact that coding editors play havoc with glyph sizing in order to fit everything into a standardized box.

> (I postulated that a system advanced enough to understand / respect the useTypoMetrics flag is also advanced enough to understand / implement the rclt feature (which should be turned on by default). And one that isn’t smart enough to implement sTypo wouldn’t implement rclt.)

> If you have any ideas of how better to implement, I’d definitely be interested! Getting the PL characters to consistently line up when you have two different metric sets is a pain.

**Comment by otherjoel** [comment]
> There are fewer powerline glyphs that I realized – I thought they might include a lot of box-drawing glyphs, but it is basically 8 useful glyphs, then a bunch of glyphs that seem to be purely for ... style?

It would be really nice to have the box-drawing glyphs, perhaps this could be useful? https://github.com/adobe-type-tools/box-drawing

**Comment by alerque** [comment]
@otherjoel I agree, but that should probably be a new issue. Box drawing glpyhs are in a different category that Powerline stuff.

**Comment by krish-r** [comment]
Hello @arrowtype, thanks for implementing this request, 

Would you like to keep this issue open or do you think I can close this issue based on 1.066 release?

![image](https://user-images.githubusercontent.com/54745129/97203539-f83d2c00-17da-11eb-96d1-66c1a4977976.png)

**Comment by arrowtype** [comment]
Thanks, @krish-r! I will keep this open for now, as I still want to implement different sizing of the power line glyphs likely to show up in certain terminals such as the integrated terminal in VS Code.

**Comment by arrowtype** [comment]
Digging into this further, it seems like this is [a VS Code issue](https://github.com/microsoft/vscode/issues/84432), which is also affecting fonts like Cascadia Code & Fira Code (and many others). In turn, this seems to maaaybe be a Chrome text rendering issue. 

The alignment is very good in the macOS Terminal and in iTerm2, so I will close this issue, and make [a new issue to investigate whether there is a reasonable fix I can do on my end for VS Code alignment](https://github.com/arrowtype/recursive/issues/421).

**macOS Terminal:**

![image](https://user-images.githubusercontent.com/45946693/99815257-0f401700-2b18-11eb-9988-5e745d7336e0.png)

**iTerm2:**

![image](https://user-images.githubusercontent.com/45946693/99815342-267f0480-2b18-11eb-8604-f226b48da7bb.png)

**Comment by aaronbell** [comment]
BTW, per some of my investigations, it appears that Windows Terminal is not turning on the `rclt` feature in the command line, but does in any content printed in terminal:
https://github.com/microsoft/cascadia-code/issues/310#issuecomment-738209291

**Comment by aaronbell** [comment]
(Also you're inspiring me to go back and improve some of those ugly glyphs :) )

**Comment by weiweihuanghuang** [comment]
Thanks for documenting another obscure font issue Stephen! I was looking up what the deal was with Powerline symbols was and didn't understand the meaning of the 'extra' set.
